18-websockets/classworks/Malkov_Aleksei/src/pages/Chat/Chat.jsx

```diff
@@ -0,0 +1,112 @@
+import React, { useEffect, useState, useRef } from 'react';
+import './Chat.css';
+import Message from './components/Message';
+import io from 'socket.io-client';
+import debounce from './debounce';
+
+const ENDPOINT = "http://localhost:3001";
+const socket = io(ENDPOINT, {transports: ['websocket']});
+socket.connect();
+
+function Chat() {
+  const defaultName = 'Chel';
+  const [message, setMessage] = useState('');
+  const [listOfMessages, setListOfMessages] = useState([]);
+  const [username, setUsername] = useState(defaultName);
+  const [typing, setTyping] = useState([]);
+  const name = useRef(null);
+
+  function changeUserName() {
+    setUsername(name.current.value);
+    socket.emit("change name", name.current.value);
+  }
+
+  useEffect(() => {
+    socket.emit("add user", defaultName);
+    return () => {
+      socket.disconnect();
+    }
+  }, []);
+
+  socket.on('reg as user', (data) => {
+    setUsername(data);
+  });
+
+  socket.on('new message', (data) => {
+    setListOfMessages([...listOfMessages, data]);
+  });
+
+  socket.on('typing', (data) => {
+    if (!typing.includes(data.username)) {
+      setTimeout(() => {
+        setTyping([...typing].filter((typer) => typer !== data.username));
+      }, 2000)
+      setTyping([...typing, data.username]);
+    }
+  });
+
+  function type(e) {
+    setMessage(e.target.value);
+    const emitter = debounce(function() {
+      socket.emit("typing", username);
+    }, 2000, true);
+    emitter();
+  }
```

Comment:
I think you should wrap the whole method `type` into `debounce`:
```javascript
const debouncedType = debounce(type, 2000, true)
```
Now your code will create a new function for every type and every function has it's own timers. I think that the problem

Reply:
socket.on() in component function body was a bad idea)) but its actually dont see updates of hooked states, so i put them in useEffects
with debounce npm and 
![image](https://user-images.githubusercontent.com/8777791/89516321-f584f900-d7e0-11ea-8684-3741f397635d.png)
now browsers at least don't crash, ty)